src/components/pages/admin/contacts/RoleEditor.vue

```diff
@@ -0,0 +1,167 @@
+    >
+      <form class="flex-initial" @submit.prevent="handleFormSubmit">
+        <div class="my-2 py-1">
+          <AppSelect
+            v-model="newRole.role"
+            label="New role"
+            :items="roleOptions"
+          />
+        </div>
+
+        <div>
+          <div class="my-2 py-1">
+            <AppLabel :label="inputT('starts.label')" required />
+            <AppRadioGroup
+              v-model="roleHasStartDate"
+              name="roleStartDate"
+              :options="[
+                [false, inputT('starts.opts.now')],
+                [true, inputT('starts.opts.schedule')],
+              ]"
+              required
+            />
+            <div v-if="roleHasStartDate" class="flex gap-2">
+              <div>
+                <AppInput v-model="newRole.startDate" type="date" required />
+              </div>
+              <div>
+                <AppInput v-model="newRole.startTime" type="time" required />
+              </div>
+            </div>
+          </div>
+
+          <div class="my-2 py-1">
+            <AppRadioGroup
+              v-model="roleHasEndDate"
+              name="roleEndDate"
+              :label="inputT('expires.label')"
+              :options="[
+                [false, inputT('expires.opts.never')],
+                [true, inputT('expires.opts.schedule')],
+              ]"
+              required
+            />
+            <div v-if="roleHasEndDate" class="flex gap-2">
+              <div>
+                <AppInput v-model="newRole.endDate" type="date" required />
+              </div>
+              <div>
+                <AppInput v-model="newRole.endTime" type="time" required />
+              </div>
+            </div>
+          </div>
+        </div>
+
+        <div class="my-3 flex py-3">
+          <AppButton type="submit" variant="primary" :loading="loading">{{
+            t('contacts.data.roles.add')
+          }}</AppButton>
+          <AppButton variant="text" class="ml-2" @click="formVisible = false">{{
+            t('form.cancel')
+          }}</AppButton>
+        </div>
+      </form>
+    </div>
+  </div>
+</template>
+
+<script lang="ts" setup>
+import AppLabel from '../../../../components/forms/AppLabel.vue';
+import AppInput from '../../../../components/forms/AppInput.vue';
+import AppButton from '../../../../components/forms/AppButton.vue';
+import AppSelect from '../../../../components/forms/AppSelect.vue';
+import AppRadioGroup from '../../../../components/forms/AppRadioGroup.vue';
+import RoleEditorItem from '../../../../components/pages/admin/contacts/RoleEditorItem.vue';
+import { GetMemberData } from '../../../../utils/api/api.interface';
+import { updateRole } from '../../../../utils/api/member';
+import { onBeforeMount, ref, reactive } from 'vue';
+import { useI18n } from 'vue-i18n';
+
+const emit = defineEmits(['update']);
+const props = defineProps<{
+  contact: GetMemberData;
```

Comment:
You probably want `GetMemberDataWith<'roles'>` instead of `GetMemberData` here, will fix the `vue-tsc` error above